static_frame/core/index_hierarchy.py

```diff
@@ -72,61 +70,200 @@
 
-
 if tp.TYPE_CHECKING:
-    from pandas import DataFrame #pylint: disable=W0611 #pragma: no cover
-    from static_frame.core.frame import Frame #pylint: disable=W0611,C0412 #pragma: no cover
-    from static_frame.core.frame import FrameGO #pylint: disable=W0611,C0412 #pragma: no cover
-    from static_frame.core.series import Series #pylint: disable=W0611,C0412 #pragma: no cover
+    from pandas import DataFrame #pylint: disable=W0611 # pragma: no cover
+    from static_frame.core.frame import Frame #pylint: disable=W0611,C0412 # pragma: no cover
+    from static_frame.core.frame import FrameGO #pylint: disable=W0611,C0412 # pragma: no cover
+    from static_frame.core.series import Series #pylint: disable=W0611,C0412 # pragma: no cover
 
 IH = tp.TypeVar('IH', bound='IndexHierarchy')
+IHGO = tp.TypeVar('IHGO', bound='IndexHierarchyGO')
+IHAsType = tp.TypeVar('IHAsType', bound='IndexHierarchyAsType')
+
+SingleLabelType = tp.Tuple[tp.Hashable, ...]
+TreeNodeT = tp.Dict[tp.Hashable, tp.Union[tp.Sequence[tp.Hashable], "TreeNodeT"]]
+
+
+def build_indexers_from_product(list_lengths: tp.Sequence[int]) -> tp.List[np.ndarray]:
+    '''
+    Creates a list of indexer arrays given a sequence of `list_lengths`
+
+    This is equivalent to: ``np.array(list(itertools.product(*(range(l) for l in list_lengths))))``
+    except it scales incredibly well.
+
+    It observes that the indexers for a product will look like this:
+
+    Example:
+
+    >>> lists = [[1, 2, 3], [4, 5, 6]]
+    >>> build_indexers_from_product([3, 3])
+    [
+        array([0, 0, 0, 1, 1, 1, 2, 2, 2]),
+        array([0, 1, 2, 0, 1, 2, 0, 1, 2]),
+    ]
+
+    >>> lists = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
+    >>> build_indexers_from_product(tuple(map(len, lists)))
+    [
+        array([0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2]),
+        array([0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1]),
+        array([0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2]),
+    ]
+    '''
+
+    padded_lengths = np.full(len(list_lengths) + 2, 1, dtype=int)
+    padded_lengths[1:-1] = list_lengths
+
+    all_group_reps = np.cumprod(padded_lengths)[:-2]
+    all_index_reps = np.cumprod(padded_lengths[::-1])[-3::-1]
+
+    result = []
 
+    for i, (group_reps, index_reps) in enumerate(zip(all_group_reps, all_index_reps)):
+        subsection = np.hstack(
+            tuple(
+                map(
+                    # Repeat each index (i.e. element) `index_reps` times
+                    functools.partial(np.tile, reps=index_reps),
+                    range(padded_lengths[i+1]),
+                )
+            )
+        )
+
+        # Repeat each section `index_reps` times
+        indexer = np.tile(subsection, reps=group_reps)
+        indexer.flags.writeable = False
+        result.append(indexer)
+
+    return result
+
+
+def blocks_to_container(blocks: tp.Iterator[np.ndarray]) -> np.ndarray:
+    return TypeBlocks.from_blocks(blocks).values
```

Comment:
I see; this is used only in the `via_*` interfaces where we expect to get back blocks of uniform type anyways, and want to return an array. Best would be to re-implement implement this without using `TypeBlocks`; it would require realizing the iterator as a list so that two passe could be performed: the first to evaluate type and final size, the second to populate a new array. This is out-of-scope perhaps of this PR, so maybe just create an issue for future enhancement.

Reply:
https://github.com/InvestmentSystems/static-frame/issues/443